util/Midi/MIDI.gd

```diff
@@ -0,0 +1,272 @@
+class_name MIDI extends Resource
+
+# MIDI classes
+class MidiEvent:
+	enum { NoteOff, NoteOn, Other }
+	var type; var key; var wall_tick; var delta_tick
+	func _init(t, k = 0, w = 0, d = 0):
+		self.type = t
+		self.key = k
+		self.wall_tick = w
+		self.delta_tick = d
+
+class MidiNote:
+	func _init(key = 0, velocity = 0, start_time = 0, duration = 0):
+		self.key = key
+		self.velocity = velocity
+		self.start_time = start_time
+		self.duration = duration
+	pass
+
+class MidiTrack:
+	var name; var instrument; var events; var notes
+	var max_note; var min_note
+	func _init(na = "", i = "", e = [],
+		no = [], mi=64, ma = 64):
+			name = na
+			instrument = i
+			events = e
+			notes = no
+			max_note = ma
+			min_note = mi
+
+# Enums
+
+# MIDI Event Names
+enum EventName  {
+	VoiceNoteOff = 0x80,
+	VoiceNoteOn = 0x90,
+	VoiceAftertouch = 0xA0,
+	VoiceControlChange = 0xB0,
+	VoiceProgramChange = 0xC0,
+	VoiceChannelPressure = 0xD0,
+	VoicePitchBend = 0xE0,
+	SystemExclusive = 0xF0,
+	}
+#Meta Event Names
+enum MetaEventName  {
+	MetaSequence = 0x00,
+	MetaText = 0x01,
+	MetaCopyright = 0x02,
+	MetaTrackName = 0x03,
+	MetaInstrumentName = 0x04,
+	MetaLyrics = 0x05,
+	MetaMarker = 0x06,
+	MetaCuePoint = 0x07,
+	MetaChannelPrefix = 0x20,
+	MetaEndOfTrack = 0x2F,
+	MetaSetTempo = 0x51,
+	MetaSMPTEOffset = 0x54,
+	MetaTimesSignature = 0x58,
+	MetaKeySignature = 0x59,
+	MetaSequencerSpecifi = 0x7F,
+	}
+
+
+# Vars
+var tracks = []
+export var tempo = 0
+export var bpm = 0
+
+
+# function to parse files to midi
+func parse_file(filename: String = "") -> bool:
+	# Open file, if any errors return false
+	var input_file = File.new()
+	var err = input_file.open(filename, File.READ)
+	if err:
+		return false
```

Comment:
Add a push_error here just in case so we know what the error is :)

Reply:
will do